packages/block-library/src/post-featured-image/edit.js

```diff
@@ -64,14 +65,33 @@ export default function PostFeaturedImageEdit( {
 		sizeSlug,
 		rel,
 		linkTarget,
+		useFirstImageFromPost,
 	} = attributes;
-	const [ featuredImage, setFeaturedImage ] = useEntityProp(
+	let [ featuredImage, setFeaturedImage ] = useEntityProp(
 		'postType',
 		postTypeSlug,
 		'featured_media',
 		postId
 	);
 
+	// Fallback to post content if no featured image is set.
+	// This is needed for the "Use first image from post" option.
+	const [ postContent ] = useEntityProp(
+		'postType',
+		postTypeSlug,
+		'content',
+		postId
+	);
+	const blocks = parse( postContent );
+	const imageBlock = blocks.find( ( block ) => block.name === 'core/image' );
```

Comment:
Thanks for this. I think the regex should be `/<!--\s+\/wp:(?:core\/)?image\s+-->/`

Reply:
Good call on that missing `\`, thank you.

Yeah @ockham and I were working on a similar problem and I think we may have some more convenient ways to do this later, but for now I think this will help a lot.

Reply:
If we're already going for a regex to find the image, may as well grab just the image instead of the whole post up until the image, right?

@dmsnell I expect you're more well versed in PHP regex performance and block parsing than I am, but I believe this should do the trick.

```
<!--\s+wp:(?:core\/)?image\s+(?:{(?:(?:[^}]+|}+(?=})|(?!}\s+\/?-->).)*+)?}\s+)?-->(?s:.*)<!--\s+\/wp:(?:core\/)?image\s+-->
```

The entire match can be passed into the parser so only the image block needs to be parsed.

Reply:
Actually, since we only care about the `id`, we probably don't even need to `parse` the entire block. The `id`, if it exists, is _always_ in the JSON attributes—it isn't sourced from the content.

```js
const imageOpener = /<!--\s+wp:(?:core\/)?image\s+(?<attrs>{(?:(?:[^}]+|}+(?=})|(?!}\s+\/?-->).)*)?}\s+)?-->/.exec( postContent );
const imageId = imageOpener && JSON.parse( imageOpener.groups.attrs )?.id;
```

I _think_ something like this should work, but I didn't try running it.

I forgot JS RegEx doesn't support the possessive qualifier used in the last example, but I don't think that's a problem if we have valid JSON, so I removed it here.

Reply:
@ajlende great questions and proposals 👏 

the reason I did choose the closing tag was ease of the regex pattern and a "good enough" spirit compared to parsing the entire post. yours is better, but if we're trying to optimize further I'd recommend these as well:

 - stay as close to the block parser regex patterns as possible
 - don't combine the regexes. find the block opener, and if necessary, start at the offset following the opener and then find the closer.
 - if all we need is the `id` and we're going to discard the image `src` then that seems fitting. the JS regex pattern found in the block default parser should be adequate; is it failing here?

all in all this is so much better performance wise than parsing the entire post. I hope that in the next couple releases we'll have a lazy block parser that mostly eliminates all the performance constraints that lead us to write specialized code like this.

Reply:
> if all we need is the id and we're going to discard the image src then that seems fitting. the JS regex pattern found in the block default parser should be adequate; is it failing here?

Maybe we can discard the check altogether and just use the block parser then?

Reply:
it's probably still worth not using the block parser, though that would be fine. in this particular case we're doing a lot of work without the need to. my comment about the lazy parser is there to show where we're going, but it's not ready yet, and that's still in PHP. sometimes I get confused thinking about what's on the server vs. what's in the browser